lib/os/onoff.c

```diff
@@ -0,0 +1,627 @@
+		srv->start(srv, onoff_start_notify);
+	} else if (notify) {
+		notify_one(srv, cli, 0);
+	}
+
+	return rv;
+}
+
+static void onoff_stop_notify(struct onoff_service *srv,
+			      int res)
+{
+	bool notify_clients = false;
+	int client_res = res;
+	bool start = false;
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+	sys_slist_t clients = srv->clients;
+	struct onoff_client *releaser = srv->releaser;
+
+	/* If the stop operation failed log an error and leave the
+	 * rest of the state in place.
+	 *
+	 * If it succeeded remove the last reference and transition to
+	 * off.
+	 *
+	 * In either case remove the last reference, and notify all
+	 * waiting clients of operation completion.
+	 */
+	if (res < 0) {
+		srv->flags &= ~SERVICE_STATE_TRANSITION;
+		srv->flags |= ONOFF_SERVICE_HAS_ERROR;
+		notify_clients = true;
+	} else if (sys_slist_is_empty(&clients)) {
+		set_service_state(srv, SERVICE_STATE_OFF);
+	} else if (k_is_in_isr()
+		   && ((srv->flags & ONOFF_SERVICE_START_WAITS) != 0U)) {
+		set_service_state(srv, SERVICE_STATE_OFF);
+		notify_clients = true;
+		client_res = -EWOULDBLOCK;
+	} else {
+		set_service_state(srv, SERVICE_STATE_TO_ON);
+		start = true;
+	}
+
+	if (releaser) {
+		srv->refs -= 1U;
+		srv->releaser = NULL;
+	}
+	__ASSERT_NO_MSG(srv->refs == 0);
+
+	/* Remove the clients if there was an error or a delayed start
+	 * couldn't be initiated, because we're resolving their
+	 * operation with an error.
+	 */
+	if (notify_clients) {
+		sys_slist_init(&srv->clients);
+	}
+
+	k_spin_unlock(&srv->lock, key);
+
+	/* Notify the releaser (if available).  If there was an error,
+	 * notify any pending requests; otherwise if there are pending
+	 * requests start the transition to ON.
+	 */
+	if (releaser) {
+		notify_one(srv, releaser, res);
+	}
+	if (notify_clients) {
+		notify_all(srv, &clients, client_res);
+	} else if (start) {
+		srv->start(srv, onoff_start_notify);
+	}
+}
+
+int z_impl_onoff_release(struct onoff_service *srv,
+			 struct onoff_client *cli)
+{
+	bool stop = false;      /* invoke stop transition */
+	bool notify = false;    /* do client notification */
+	bool no_wait = false;
+	int rv = validate_args(srv, cli, &no_wait);
+
+	if (rv < 0) {
+		return rv;
+	}
+
+	if (rv == ONOFF_CLIENT_NOTIFY_ANONYMOUS) {
+		return -ENOTSUP;
+	}
+
+	rv = 0;
+
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+
+	if ((srv->flags & ONOFF_SERVICE_HAS_ERROR) != 0) {
+		rv = -EIO;
+		goto out;
+	}
+
+	u32_t state = srv->flags & SERVICE_STATE_MASK;
+
+	switch (state) {
+	case SERVICE_STATE_ON:
+		/* Stay on if release leaves a client. */
+		if (srv->refs > 1U) {
+			notify = true;
+			rv = 1;
+			break;
+		}
+
+		/* Reject if in no-wait context but stop could
+		 * wait.
+		 */
+		if ((no_wait || k_is_in_isr())
+		    && ((srv->flags & ONOFF_SERVICE_STOP_WAITS) != 0)) {
+			rv = -EWOULDBLOCK;
+			break;
+		}
+
+		stop = true;
+
+		set_service_state(srv, SERVICE_STATE_TO_OFF);
+		srv->releaser = cli;
```

Comment:
can you add a flag to client which would prevent being assigned as a releaser? In case client is not interested in getting notified that he is a releaser.

Reply:
The state of this field also affects management of the reference count.